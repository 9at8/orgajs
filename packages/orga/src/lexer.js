const { escape } = require('./utils')
const timestamp = require('./timestamp')

function Syntax() {
  this.rules = []
}

Syntax.prototype = {
  define: function(name, pattern, post = () => { return {} }) {
    this.rules.push({
      name,
      pattern,
      post,
    })
  },

  update: function(name, pattern) {
    const i = this.rules.findIndex(r => r.name === name)
    var newRule = { name, post: () => {} }
    if (i !== -1) {
      newRule = this.rules.splice(i, 1)[0]
    }
    newRule.pattern = pattern
    this.rules.splice(i, 0, newRule)
  }
}

var org = new Syntax()

function headlinePattern(todos = ['TODO', 'DONE']) {
  return RegExp(`^(\\*+)\\s+(?:(${todos.map(escape).join('|')})\\s+)?(?:\\[#(A|B|C)\\]\\s+)?(.*?)\\s*(:(?:\\w+:)+)?$`)
}

org.define('headline', headlinePattern(), m => {
  const level = m[1].length
  const keyword = m[2]
  const priority = m[3]
  const content = m[4]
  const tags = (m[5] || '').split(':').map( str => str.trim()).filter(String)
  return { level, keyword, priority, content, tags }
})

org.define('keyword', /^\s*#\+(\w+):\s*(.*)$/, m => {
  const key = m[1]
  const value = m[2]
  return { key, value }
})

const PLANNING_KEYWORDS = ['DEADLINE', 'SCHEDULED', 'CLOSED']
org.define('planning', RegExp(`^\\s*(${PLANNING_KEYWORDS.join('|')}):\\s*(.+)$`), m => {
  const keyword = m[1]
  return { keyword, ...timestamp.parse(m[2]) }
})

org.define('timestamp', RegExp(timestamp.pattern, 'i'), m => {
  return timestamp.parse(m)
})

org.define('block.begin', /^\s*#\+begin_(\w+)(.*)$/i, m => {
  const type = m[1]
  const params = m[2].split(' ').map( str => str.trim()).filter(String)
  return { type, params }
})

org.define('block.end', /^\s*#\+end_(\w+)$/i, m => {
  const type = m[1]
  return { type }
})

org.define('drawer.end', /^\s*:end:\s*$/i)

org.define('drawer.begin', /^\s*:(\w+):\s*$/, m => {
  const type = m[1]
  return { type }
})

org.define('list.item', /^(\s*)([-+]|\d+[.)])\s+(?:\[(x|X|-| )\][ \t]+)?(?:([^\n]+)[ \t]+::[ \t]*)?(.*)$/, m => {
  const indent = m[1].length
  const bullet = m[2]
  const tag = m[4]
  const content = m[5]

  var ordered = true
  if ( [`-`, `+`].includes(bullet) ) {
    ordered = false
  }

  var result = { indent, ordered, content, tag }
  if (m[3]) {
    var checked = m[3] !== ' '
    result.checked = checked
  }
  return result
})

org.define('table.separator', /^\s*\|-/)

org.define('table.row', /^\s*\|(\s*.+\|)+\s*$/, m => {
  const cells = m[1].split('|').map( str => str.trim()).filter(String)
  return { cells }
})

org.define('horizontalRule', /^\s*-{5,}\s*$/)

org.define('comment', /^\s*#\s.*$/)

org.define('footnote', /^\[fn:(\w+)\]\s+(.*)$/, m => {
  const label = m[1]
  const content = m[2]
  return { label, content }
})

function Lexer(options = require('./defaults')) {
  this.syntax = org
  const { todos } = options
  if (todos) {
    this.updateTODOs(todos)
  }
}

Lexer.prototype = {
  tokenize: function (input) {
    for ( const { name, pattern, post } of this.syntax.rules ) {
      const m = pattern.exec(input)
      if (!m) { continue }
      var token = { name, raw: input }
      token.data = post(m)
      return token
    }

    const trimed = input.trim()
    if (trimed === '') {
      return { name: `blank`, raw: input }
    }

    return { name: `line`, raw: input }
  },

  updateTODOs: function(todos) {
    this.syntax.update(`headline`, headlinePattern(todos))
  }

}

module.exports = Lexer
