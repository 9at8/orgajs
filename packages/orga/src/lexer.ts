import defaultOptions, { ParseOptions } from './options'
import { escape } from './utils'
import { parse as parseTimestamp, pattern as timestampPattern } from './timestamp'
import XRegExp from 'xregexp'
import { read } from './reader'
import { tokenize as inlineTok } from './tokenize/inline'
import tokenizeHeadline from './tokenize/headline'
import tokenizePlanning from './tokenize/planning'
import tokenizeBlock from './tokenize/block'
import tokenizeListItem from './tokenize/list'
import tokenizeFootnote from './tokenize/footnote'
import tokenizeDrawer from './tokenize/drawer'
import { isEmpty } from './position'

type Rule = {
  name: string
  pattern: RegExp
  post: (m: any, options?: ParseOptions) => void
}

export interface Lexer {
  next: () => Token | undefined;
  peek: () => Token | undefined;
  all: () => Token[];
}

class Syntax {
  rules: Rule[]

  constructor() {
    this.rules = []
  }

  define(name: string,
         pattern: RegExp,
         post = (_: any, _opitons?: ParseOptions) => { return {} }) : void {
    this.rules.push({
      name,
      pattern,
      post,
    })
  }

  update(name: string, pattern: RegExp) {
    // const i = this.rules.findIndex(r => r.name === name)
    // let newRule = { name, post: (_: any) => {}, pattern: undefined }
    // if (i !== -1) {
    //   newRule = this.rules.splice(i, 1)[0]

    // }
    // newRule.pattern = pattern
    // this.rules.splice(i, 0, newRule)
  }
}

const org = new Syntax()

function headlinePattern(todos = ['TODO', 'DONE']) {
  return RegExp(`^(\\*+)\\s+(?:(${todos.map(escape).join('|')})\\s+)?(?:\\[#(A|B|C)\\]\\s+)?(.*?)\\s*(:(?:[\\w@]+:)+)?$`)
}

org.define('headline', headlinePattern(), m => {
  const level = m[1].length
  const keyword = m[2]
  const priority = m[3]
  const content = m[4]
  const tags = (m[5] || '').split(':').map((str: string) => str.trim()).filter(String)
  return { level, keyword, priority, content, tags }
})

org.define('keyword', /^\s*#\+(\w+):\s*(.*)$/, m => {
  const key = m[1]
  const value = m[2]
  return { key, value }
})

const PLANNING_KEYWORDS = ['DEADLINE', 'SCHEDULED', 'CLOSED']
org.define('planning', RegExp(`^\\s*(${PLANNING_KEYWORDS.join('|')}):\\s*(.+)$`), (m, options) => {
  const keyword = m[1]
  return { keyword, ...parseTimestamp(m[2], options) }
})

// org.define('timestamp', XRegExp(timestampPattern, 'i'), (m, options) => {
//   // console.log(options)
//   return parseTimestamp(m, options)
// })

org.define('block.begin', /^\s*#\+begin_(\w+)(.*)$/i, m => {
  const type = m[1]
  const params = m[2].split(' ').map( str => str.trim()).filter(String)
  return { type, params }
})

org.define('block.end', /^\s*#\+end_(\w+)$/i, m => {
  const type = m[1]
  return { type }
})

org.define('drawer.end', /^\s*:end:\s*$/i)

org.define('drawer.begin', /^\s*:(\w+):\s*$/, m => {
  const type = m[1]
  return { type }
})

org.define('list.item', /^(\s*)([-+]|\d+[.)])\s+(?:\[(x|X|-| )\][ \t]+)?(?:([^\n]+)[ \t]+::[ \t]*)?(.*)$/, m => {
  const indent = m[1].length
  const bullet = m[2]
  const tag = m[4]
  const content = m[5]

  let ordered = true
  if ( [`-`, `+`].includes(bullet) ) {
    ordered = false
  }

  const result = {
    indent,
    ordered,
    content,
    tag,
    checked: false }
  if (m[3]) {
    const checked = m[3] !== ' '
    result.checked = checked
  }
  return result
})

org.define('table.separator', /^\s*\|-/)

org.define('table.row', /^\s*\|([^-].*\|.*)\|[ \t]*$/, m => {
  const cells = m[1].split('|').map( str => str.trim())
  return { cells }
})

org.define('horizontalRule', /^\s*-{5,}\s*$/)

org.define('comment', /^\s*#\s.*$/)

org.define('footnote', /^\[fn:(\w+)\]\s+(.*)$/, m => {
  const label = m[1]
  const content = m[2]
  return { label, content }
})

export default class OldLexer {
  syntax: any
  options: ParseOptions

  constructor(options: ParseOptions) {
    this.syntax = org
    this.options = { ...defaultOptions, ...options }
    const { todos } = this.options
    if (todos) {
      this.updateTODOs(todos)
    }
  }

  tokenize(input: string): any {
    for ( const { name, pattern, post } of this.syntax.rules ) {
      const m = pattern.exec(input)
      if (!m) { continue }
      return {
        name,
        raw: input,
        data: post(m, this.options) }
    }

    const trimed = input.trim()
    if (trimed === '') {
      return { name: `blank`, raw: input }
    }

    return { name: `line`, raw: input }
  }

  updateTODOs(todos: string[]) {
    this.syntax.update(`headline`, headlinePattern(todos))
  }
}

export const tokenize = (text: string, options: ParseOptions = defaultOptions) => {

  const { timezone, todos } = { ...defaultOptions, ...options }

  const reader = read(text)

  const {
    isStartOfLine,
    eat,
    getLine,
    getChar,
    now,
    match,
    EOF,
    skipWhitespaces,
  } = reader

  let todoKeywords = todos

  let buffer: Token[] = []

  const tok = (): Token[] => {
    skipWhitespaces()
    if (EOF()) return []

    if (getChar() === '\n') {
      return [{
        name: 'newline',
        position: eat('char'),
      }]
    }

    if (isStartOfLine() && match(/^\*+\s+/)) {
      return tokenizeHeadline({ reader, todoKeywords })
    }

    const l = getLine()
    if (PLANNING_KEYWORDS.some((k) => l.startsWith(k))) {
      return tokenizePlanning({
        reader,
        keywords: PLANNING_KEYWORDS,
        timezone })
    }

    if (l.startsWith('#+')) {

      const keyword = match(/^\s*#\+(\w+):\s*(.*)$/)
      if (keyword) {
        eat('line')
        return [{
          name: 'keyword',
          data: { key: keyword.captures[1], value: keyword.captures[2] },
          position: keyword.position,
        }]
      }

      const block = tokenizeBlock({ reader })
      if (block.length > 0) return block
    }

    const list = tokenizeListItem({ reader })
    if (list.length > 0) return list

    if (l.startsWith('# ')) {
      const comment = eat(/^#\s.*$/)
      if (!isEmpty(comment)) {
        return [ { name: 'comment', position: comment } ]
      }
    }

    const drawer = tokenizeDrawer({ reader })
    if (drawer.length > 0) return drawer
    // TODO: table

    const hr = eat(/^\s*-{5,}\s*$/)
    if (!isEmpty(hr)) {
      return [{
        name: 'hr',
        position: hr,
      }]
    }

    if (now().column === 0) {
      const footnote = tokenizeFootnote({ reader })
      if (footnote.length > 0) return footnote
    }

    // last resort
    return inlineTok({ reader })
  }

  const peek = () : Token | undefined => {
    if (buffer.length === 0) {
      buffer = tok()
    }
    return buffer[0]
  }


  const next = () : Token | undefined => {
    if (buffer.length === 0) {
      buffer = tok()
    }
    return buffer.shift()
  }

  const all = (max: number | undefined = undefined) : Token[] => {
    const tokens: Token[] = []
    let token
    do {
      token = next()
      if (token) {
        tokens.push(token)
      }
      if (max && tokens.length >= max) {
        break
      }
    } while(token)
    return tokens
  }

  return {
    next,
    peek,
    all,
  } as Lexer
}

