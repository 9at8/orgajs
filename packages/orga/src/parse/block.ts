import { Lexer } from '../tokenize'
import { newNode, push, Node } from '../node'

export default (lexer: Lexer): Node | undefined => {

  const { peek, eat } = lexer

  const n = peek()

  if (!n || n.type !== 'block.begin') return undefined

  const block = newNode('block')
  block.data = { ...n.data }
  const a = push(block)
  a(n)
  eat()

  const parse = (): Node | undefined => {
    const n = peek()
    if (!n || n.type === 'stars') return undefined
    a(n)
    eat()
    if (n.type === 'block.end') {
      eat('newline')
      return block
    } else {
      return parse()
    }
  }

  return parse()
}
