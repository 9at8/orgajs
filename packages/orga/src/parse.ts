import { Lexer } from './lexer'
import { newNode, push, Node } from './node'
import { inspect } from 'util'

export const parse = ({ next, peek }: Lexer) => {
  const tree = newNode('document')

  const collectContent = (container: Node): Node => {
    const a = push(container)
    const token = peek()
    // if (!token) return container
    // if (token.name === 'newline') return container
    if (!token || !token.name.startsWith('text.')) return container
    a(token)
    next()
    return collectContent(container)
  }

  const parseHeadline = (headline: Node = newNode('headline')): Node => {
    const a = push(headline)

    const token = peek()
    if (!token) return headline
    if (token.name === 'newline') {
      next()
      return headline
    }

    if (['stars', 'keyword', 'priority'].includes(token.name)) {
      a(token)
      next()
      return parseHeadline(headline)
    }

    const content = collectContent(newNode('content'))
    if (content.children.length > 0) {
      a(content)
      return parseHeadline(headline)
    }

    next()
    return parseHeadline(headline)
  }

  const parsePlanning = (section: Node): Node => {
    const token = peek()
    if (!token || token.name !== 'planning.keyword') return section
    const planning = newNode('planning')
    const collect = push(planning)
    collect(token)
    next()
    const timestamp = peek()
    if (timestamp && timestamp.name === 'planning.timestamp') {
      collect(timestamp)
      next()
    }

    push(section)(planning)
    return parsePlanning(section)
  }

  const parseSection = (): Node => {
    const section = newNode('section')
    push(section)(parseHeadline())
    parsePlanning(section)
    return section
  }

  const main = () => {
    const token = peek()
    if (!token) return
    if (token.name === 'stars') {
      push(tree)(parseSection())
    }

    // lose the token
    console.log('discard', { token: token.name })
    next()

    main()
  }

  main()

  return tree
}
