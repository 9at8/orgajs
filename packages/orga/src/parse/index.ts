import { Lexer } from '../lexer'
import { newNode, push, Node } from '../node'
import { inspect } from 'util'
import parseList from './list'
import section from './section'

export const parse = (lexer: Lexer) => {
  const { next, peek } = lexer
  const tree = newNode('document')

  const main = () => {
    const token = peek()
    if (!token) return
    if (token.type === 'stars') {
      push(tree)(section(lexer))
      main()
      return
    }

    if (token.type === 'list.item.bullet') {
      const list = parseList(lexer)
      push(tree)(list)
    }

    // lose the token
    console.log('discard', { token: token.type })
    next()

    main()
  }

  main()

  return tree
}
