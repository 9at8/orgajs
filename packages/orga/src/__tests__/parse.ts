import { parse } from '../parse'
import { tokenize } from '../lexer'
import { map } from '../position'
import { inspect } from 'util'

const debug = (text: string, tree) => {
  console.log(inspect(tree, false, null, true))
}

describe('Parser', () => {
  it('works', () => {
    const content = `
#+TITLE: hello world
#+TODO: TODO NEXT | DONE
#+DATE: 2018-01-01

* TODO [#A] headline one
DEADLINE: <2018-01-01 Mon>
:PROPERTIES:
key: value
key: value
:END:

[[https://github.com/xiaoxinghu/orgajs][Here's]] to the *crazy* ones, the /misfits/, the _rebels_, the ~troublemakers~,
the round pegs in the +round+ square holes...
`
    // const thing = parse(tokenize(content))
    const lexer = tokenize(content)

    const tree = parse(lexer)
    debug(content, tree)
    // debug(content, thing)
  })
})
