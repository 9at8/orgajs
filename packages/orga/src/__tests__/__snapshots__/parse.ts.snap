// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser can handle nested headlines 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "data": Object {
                "level": 1,
              },
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 1,
                },
                "position": Object {
                  "end": Object {
                    "column": 14,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 1,
                  },
                },
                "type": "headline",
              },
              "position": Object {
                "end": Object {
                  "column": 1,
                  "line": 1,
                },
                "start": Object {
                  "column": 0,
                  "line": 1,
                },
              },
              "type": "stars",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": Object {
                    "children": [Circular],
                    "position": Object {
                      "end": Object {
                        "column": 14,
                        "line": 1,
                      },
                      "start": Object {
                        "column": 2,
                        "line": 1,
                      },
                    },
                    "type": "content",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 14,
                      "line": 1,
                    },
                    "start": Object {
                      "column": 2,
                      "line": 1,
                    },
                  },
                  "type": "text.plain",
                },
              ],
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 1,
                },
                "position": Object {
                  "end": Object {
                    "column": 14,
                    "line": 1,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 1,
                  },
                },
                "type": "headline",
              },
              "position": Object {
                "end": Object {
                  "column": 14,
                  "line": 1,
                },
                "start": Object {
                  "column": 2,
                  "line": 1,
                },
              },
              "type": "content",
            },
          ],
          "data": Object {
            "level": 1,
          },
          "parent": Object {
            "children": [Circular],
            "data": Object {
              "level": 1,
            },
            "position": Object {
              "end": Object {
                "column": 17,
                "line": 7,
              },
              "start": Object {
                "column": 0,
                "line": 1,
              },
            },
            "type": "section",
          },
          "position": Object {
            "end": Object {
              "column": 14,
              "line": 1,
            },
            "start": Object {
              "column": 0,
              "line": 1,
            },
          },
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "parent": Object {
                "children": [Circular],
                "position": Object {
                  "end": Object {
                    "column": 9,
                    "line": 2,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 2,
                  },
                },
                "type": "paragraph",
              },
              "position": Object {
                "end": Object {
                  "column": 9,
                  "line": 2,
                },
                "start": Object {
                  "column": 0,
                  "line": 2,
                },
              },
              "type": "text.plain",
            },
          ],
          "parent": Object {
            "children": [Circular],
            "data": Object {
              "level": 1,
            },
            "position": Object {
              "end": Object {
                "column": 17,
                "line": 7,
              },
              "start": Object {
                "column": 0,
                "line": 1,
              },
            },
            "type": "section",
          },
          "position": Object {
            "end": Object {
              "column": 9,
              "line": 2,
            },
            "start": Object {
              "column": 0,
              "line": 2,
            },
          },
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "data": Object {
                    "level": 2,
                  },
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 2,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 17,
                        "line": 3,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 3,
                      },
                    },
                    "type": "headline",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 2,
                      "line": 3,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 3,
                    },
                  },
                  "type": "stars",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": Object {
                        "children": [Circular],
                        "position": Object {
                          "end": Object {
                            "column": 17,
                            "line": 3,
                          },
                          "start": Object {
                            "column": 3,
                            "line": 3,
                          },
                        },
                        "type": "content",
                      },
                      "position": Object {
                        "end": Object {
                          "column": 17,
                          "line": 3,
                        },
                        "start": Object {
                          "column": 3,
                          "line": 3,
                        },
                      },
                      "type": "text.plain",
                    },
                  ],
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 2,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 17,
                        "line": 3,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 3,
                      },
                    },
                    "type": "headline",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 17,
                      "line": 3,
                    },
                    "start": Object {
                      "column": 3,
                      "line": 3,
                    },
                  },
                  "type": "content",
                },
              ],
              "data": Object {
                "level": 2,
              },
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 2,
                },
                "position": Object {
                  "end": Object {
                    "column": 7,
                    "line": 5,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 3,
                  },
                },
                "type": "section",
              },
              "position": Object {
                "end": Object {
                  "column": 17,
                  "line": 3,
                },
                "start": Object {
                  "column": 0,
                  "line": 3,
                },
              },
              "type": "headline",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "data": Object {
                        "level": 3,
                      },
                      "parent": Object {
                        "children": [Circular],
                        "data": Object {
                          "level": 3,
                        },
                        "position": Object {
                          "end": Object {
                            "column": 20,
                            "line": 4,
                          },
                          "start": Object {
                            "column": 0,
                            "line": 4,
                          },
                        },
                        "type": "headline",
                      },
                      "position": Object {
                        "end": Object {
                          "column": 3,
                          "line": 4,
                        },
                        "start": Object {
                          "column": 0,
                          "line": 4,
                        },
                      },
                      "type": "stars",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [],
                          "parent": Object {
                            "children": [Circular],
                            "position": Object {
                              "end": Object {
                                "column": 20,
                                "line": 4,
                              },
                              "start": Object {
                                "column": 4,
                                "line": 4,
                              },
                            },
                            "type": "content",
                          },
                          "position": Object {
                            "end": Object {
                              "column": 20,
                              "line": 4,
                            },
                            "start": Object {
                              "column": 4,
                              "line": 4,
                            },
                          },
                          "type": "text.plain",
                        },
                      ],
                      "parent": Object {
                        "children": [Circular],
                        "data": Object {
                          "level": 3,
                        },
                        "position": Object {
                          "end": Object {
                            "column": 20,
                            "line": 4,
                          },
                          "start": Object {
                            "column": 0,
                            "line": 4,
                          },
                        },
                        "type": "headline",
                      },
                      "position": Object {
                        "end": Object {
                          "column": 20,
                          "line": 4,
                        },
                        "start": Object {
                          "column": 4,
                          "line": 4,
                        },
                      },
                      "type": "content",
                    },
                  ],
                  "data": Object {
                    "level": 3,
                  },
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 3,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 7,
                        "line": 5,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 4,
                      },
                    },
                    "type": "section",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 20,
                      "line": 4,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 4,
                    },
                  },
                  "type": "headline",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": Object {
                        "children": [Circular],
                        "position": Object {
                          "end": Object {
                            "column": 7,
                            "line": 5,
                          },
                          "start": Object {
                            "column": 0,
                            "line": 5,
                          },
                        },
                        "type": "paragraph",
                      },
                      "position": Object {
                        "end": Object {
                          "column": 7,
                          "line": 5,
                        },
                        "start": Object {
                          "column": 0,
                          "line": 5,
                        },
                      },
                      "type": "text.plain",
                    },
                  ],
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 3,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 7,
                        "line": 5,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 4,
                      },
                    },
                    "type": "section",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 7,
                      "line": 5,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 5,
                    },
                  },
                  "type": "paragraph",
                },
              ],
              "data": Object {
                "level": 3,
              },
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 2,
                },
                "position": Object {
                  "end": Object {
                    "column": 7,
                    "line": 5,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 3,
                  },
                },
                "type": "section",
              },
              "position": Object {
                "end": Object {
                  "column": 7,
                  "line": 5,
                },
                "start": Object {
                  "column": 0,
                  "line": 4,
                },
              },
              "type": "section",
            },
          ],
          "data": Object {
            "level": 2,
          },
          "parent": Object {
            "children": [Circular],
            "data": Object {
              "level": 1,
            },
            "position": Object {
              "end": Object {
                "column": 17,
                "line": 7,
              },
              "start": Object {
                "column": 0,
                "line": 1,
              },
            },
            "type": "section",
          },
          "position": Object {
            "end": Object {
              "column": 7,
              "line": 5,
            },
            "start": Object {
              "column": 0,
              "line": 3,
            },
          },
          "type": "section",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "data": Object {
                    "level": 2,
                  },
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 2,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 17,
                        "line": 7,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 7,
                      },
                    },
                    "type": "headline",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 2,
                      "line": 7,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 7,
                    },
                  },
                  "type": "stars",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": Object {
                        "children": [Circular],
                        "position": Object {
                          "end": Object {
                            "column": 17,
                            "line": 7,
                          },
                          "start": Object {
                            "column": 3,
                            "line": 7,
                          },
                        },
                        "type": "content",
                      },
                      "position": Object {
                        "end": Object {
                          "column": 17,
                          "line": 7,
                        },
                        "start": Object {
                          "column": 3,
                          "line": 7,
                        },
                      },
                      "type": "text.plain",
                    },
                  ],
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 2,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 17,
                        "line": 7,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 7,
                      },
                    },
                    "type": "headline",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 17,
                      "line": 7,
                    },
                    "start": Object {
                      "column": 3,
                      "line": 7,
                    },
                  },
                  "type": "content",
                },
              ],
              "data": Object {
                "level": 2,
              },
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 2,
                },
                "position": Object {
                  "end": Object {
                    "column": 17,
                    "line": 7,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 7,
                  },
                },
                "type": "section",
              },
              "position": Object {
                "end": Object {
                  "column": 17,
                  "line": 7,
                },
                "start": Object {
                  "column": 0,
                  "line": 7,
                },
              },
              "type": "headline",
            },
          ],
          "data": Object {
            "level": 2,
          },
          "parent": Object {
            "children": [Circular],
            "data": Object {
              "level": 1,
            },
            "position": Object {
              "end": Object {
                "column": 17,
                "line": 7,
              },
              "start": Object {
                "column": 0,
                "line": 1,
              },
            },
            "type": "section",
          },
          "position": Object {
            "end": Object {
              "column": 17,
              "line": 7,
            },
            "start": Object {
              "column": 0,
              "line": 7,
            },
          },
          "type": "section",
        },
      ],
      "data": Object {
        "level": 1,
      },
      "parent": [Circular],
      "position": Object {
        "end": Object {
          "column": 17,
          "line": 7,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "type": "section",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "data": Object {
                "level": 1,
              },
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 1,
                },
                "position": Object {
                  "end": Object {
                    "column": 14,
                    "line": 8,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 8,
                  },
                },
                "type": "headline",
              },
              "position": Object {
                "end": Object {
                  "column": 1,
                  "line": 8,
                },
                "start": Object {
                  "column": 0,
                  "line": 8,
                },
              },
              "type": "stars",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "parent": Object {
                    "children": [Circular],
                    "position": Object {
                      "end": Object {
                        "column": 14,
                        "line": 8,
                      },
                      "start": Object {
                        "column": 2,
                        "line": 8,
                      },
                    },
                    "type": "content",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 14,
                      "line": 8,
                    },
                    "start": Object {
                      "column": 2,
                      "line": 8,
                    },
                  },
                  "type": "text.plain",
                },
              ],
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 1,
                },
                "position": Object {
                  "end": Object {
                    "column": 14,
                    "line": 8,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 8,
                  },
                },
                "type": "headline",
              },
              "position": Object {
                "end": Object {
                  "column": 14,
                  "line": 8,
                },
                "start": Object {
                  "column": 2,
                  "line": 8,
                },
              },
              "type": "content",
            },
          ],
          "data": Object {
            "level": 1,
          },
          "parent": Object {
            "children": [Circular],
            "data": Object {
              "level": 1,
            },
            "position": Object {
              "end": Object {
                "column": 17,
                "line": 9,
              },
              "start": Object {
                "column": 0,
                "line": 8,
              },
            },
            "type": "section",
          },
          "position": Object {
            "end": Object {
              "column": 14,
              "line": 8,
            },
            "start": Object {
              "column": 0,
              "line": 8,
            },
          },
          "type": "headline",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "data": Object {
                    "level": 2,
                  },
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 2,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 17,
                        "line": 9,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 9,
                      },
                    },
                    "type": "headline",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 2,
                      "line": 9,
                    },
                    "start": Object {
                      "column": 0,
                      "line": 9,
                    },
                  },
                  "type": "stars",
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "parent": Object {
                        "children": [Circular],
                        "position": Object {
                          "end": Object {
                            "column": 17,
                            "line": 9,
                          },
                          "start": Object {
                            "column": 3,
                            "line": 9,
                          },
                        },
                        "type": "content",
                      },
                      "position": Object {
                        "end": Object {
                          "column": 17,
                          "line": 9,
                        },
                        "start": Object {
                          "column": 3,
                          "line": 9,
                        },
                      },
                      "type": "text.plain",
                    },
                  ],
                  "parent": Object {
                    "children": [Circular],
                    "data": Object {
                      "level": 2,
                    },
                    "position": Object {
                      "end": Object {
                        "column": 17,
                        "line": 9,
                      },
                      "start": Object {
                        "column": 0,
                        "line": 9,
                      },
                    },
                    "type": "headline",
                  },
                  "position": Object {
                    "end": Object {
                      "column": 17,
                      "line": 9,
                    },
                    "start": Object {
                      "column": 3,
                      "line": 9,
                    },
                  },
                  "type": "content",
                },
              ],
              "data": Object {
                "level": 2,
              },
              "parent": Object {
                "children": [Circular],
                "data": Object {
                  "level": 2,
                },
                "position": Object {
                  "end": Object {
                    "column": 17,
                    "line": 9,
                  },
                  "start": Object {
                    "column": 0,
                    "line": 9,
                  },
                },
                "type": "section",
              },
              "position": Object {
                "end": Object {
                  "column": 17,
                  "line": 9,
                },
                "start": Object {
                  "column": 0,
                  "line": 9,
                },
              },
              "type": "headline",
            },
          ],
          "data": Object {
            "level": 2,
          },
          "parent": Object {
            "children": [Circular],
            "data": Object {
              "level": 1,
            },
            "position": Object {
              "end": Object {
                "column": 17,
                "line": 9,
              },
              "start": Object {
                "column": 0,
                "line": 8,
              },
            },
            "type": "section",
          },
          "position": Object {
            "end": Object {
              "column": 17,
              "line": 9,
            },
            "start": Object {
              "column": 0,
              "line": 9,
            },
          },
          "type": "section",
        },
      ],
      "data": Object {
        "level": 1,
      },
      "parent": [Circular],
      "position": Object {
        "end": Object {
          "column": 17,
          "line": 9,
        },
        "start": Object {
          "column": 0,
          "line": 8,
        },
      },
      "type": "section",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 17,
      "line": 9,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "type": "document",
}
`;
